### ISSUE TITLE
Ideas for `flake8-trio` checks

### ISSUE THREAD

**Issue by Zac-HD** [Zac-HD/flake8-trio#1]
As a "list of ideas" issue, any of the ideas below might turn out to be too difficult to implement, or in fact counterproductive if implemented.  Nonetheless, we need somewhere to dump ideas and this issue is that place.  Items are in rough priority order by a combination of estimated importance and tractability.

- [x] [Never `yield` inside a nursery or cancel scope](https://github.com/python-trio/trio/issues/671#issuecomment-431646729)
  #4 
- [x] Anything that `await`s inside a `finally:` clause needs shielding *and* a timeout, e.g. `with trio.fail_after(1, shield=True):`  ([source](https://github.com/python-trio/trio/issues/671#issuecomment-578984615)) (also applies to `async for` or `async with`)
  #5
- [x] Never have an `except Cancelled` or `except BaseException` block with a code path that doesn't re-`raise` the error.
  #8
- [x] Calling a Trio async function without immediately `await`ing it.  Per [the discussion here](https://github.com/python-trio/trio/issues/671#issue-360692109) typecheckers have some support for the (difficult!) general case, but it's easy and useful to make a list of all Trio functions and check for e.g. `trio.sleep(0)` without an `await`.  We *don't* need to check `async for` or `async with` because using the sync syntax is a noisy error at runtime, unlike the checkpoint warnings.
  #10
- [x] Importing things from the main `trio` namespace in a way that isn't just `import trio` (this is really just to make the linter easier to implement, but it's consistent with ~99% of existing usage anyway)
  #11
- [x] Async functions must have at least one [checkpoint (i.e. `await`, `async for`, or `async with`)](https://trio.readthedocs.io/en/stable/reference-core.html#checkpoints) on every code path, unless an exception is raised (the same rule that Trio internals follow)
  #16
- [x] Async *iterables* (i.e. with a `yield`) should have at least one checkpoint in each loop iteration, and an additional checkpoint after the last iteration (unless an exception is raised, again like the Trio internals)
  #17
- [x] Expanding `TRIO102` - you need a shielded cancel scope to `await` inside an `except Cancelled/BaseException` block, not *just* `finally:`
  #18
- [x] Async function definitions should not have a `timeout` parameter - encourage use of `trio.fail_after()` or `trio.move_on_after()` instead.  ([`httpx` does OK here](https://www.python-httpx.org/advanced/#fine-tuning-the-configuration), but most uses are IMO a design mistake)
  #20
- [x] Prefer waiting on a `trio.Event` rather than sleeping in a loop.  Concretely I think matching `while not <expr>: await trio.sleep(...)` might work well here?
  #21
- [x] [When nesting async context managers, nurseries should be ~~outermost~~ innermost.](https://github.com/python-trio/trio/issues/671#issuecomment-533978268) (actually pretty complicated?)
  #22
- [x] [Warn if we see a nursery body that contains exactly one statement, and that statement is a call to `nursery.start[_soon]`](https://github.com/python-trio/trio/issues/671#issuecomment-534266031) ("consider using a regular function call instead"), *unless* `nursery` is passed as an argument to the start-soon'd function.
  #34

**Comment by jakkdl** [comment]
For #1, https://github.com/python-trio/trio/issues/264 has a `trio.open_cancel_scope` which I can't find in the docs. So I'm guessing that's removed, and as such we only care about yielding in nurseries?
Must this include a check that it's not within a `@contextmanager` or an `@asynccontextmanager`?

**Comment by Zac-HD** [comment]
Nurseries, and `trio.{fail,move_on}_{at,after}()`, yeah.  And yep, we'll want to check for the presence of a context-manager decorator to avoid false alarms 👍

**Comment by jakkdl** [comment]
Possibly support `import trio as <foo>` and `from trio import *` with varying levels of sophistication (assume everything called e.g. `open_nursery` is from trio, vs parse the imports) depending on need.

**Comment by jakkdl** [comment]
> Async functions should not have a timeout parameter - use trio.fail_after() or trio.move_on_after() instead. ([httpx does OK here](https://www.python-httpx.org/advanced/#fine-tuning-the-configuration), but most uses are IMO a design mistake)

Is this about looking at the function signature
```python
async def foo(timeout): # error
  ...
```
or function calls with a timeout keyword? (excluding calls to the trio library)
```python
foo(timeout=15) # error
```
or both?

**Comment by jakkdl** [comment]
> Prefer waiting on a trio.Event rather than sleeping in a loop. Concretely I think matching `while not <expr>: await trio.sleep(...)` might work well here?
is the `not` in the loop condition significant? Or should any calls to `await trio.sleep(...)` inside loops be errors?
the latter feels like it might send a lot of false alarms?
```python
for i in range(...):
  ...
  if ...:
    await trio.sleep(...)
  ...
```

**Comment by Zac-HD** [comment]
> Is this about looking at the function signature [...] or function calls with a timeout keyword?

Function *definitions* - nothing wrong with using the API available to you, but for user-controlled code Trio's idiomatic cancellation scopes are so much nicer.

> > Prefer waiting on a trio.Event rather than sleeping in a loop. Concretely I think matching `while not <expr>: await trio.sleep(...)` might work well here?
> 
> is the `not` in the loop condition significant? Or should any calls to `await trio.sleep(...)` inside loops be errors? the latter feels like it might send a lot of false alarms?

Let's go with "any `while` loop where the body consists entirely of `await trio.sleep[_until](...)`".  It's a pretty narrow check that way, but it's part of a useful category I think of as 'docs presented as you need them'.

**Comment by jakkdl** [comment]
> Async iterables (i.e. with a yield) should have at least one checkpoint in each loop iteration, and an additional checkpoint after the last iteration (unless an exception is raised, again like the Trio internals)

Should there be a checkpoint before the first yield?

**Comment by Zac-HD** [comment]
That's allowed but not required.

**Comment by jakkdl** [comment]
Would be nice if you made a comment here (or somewhere) when updating the task list - I get no notification when you edit it otherwise.

**Comment by Zac-HD** [comment]
Oops, will do!

**Comment by jakkdl** [comment]
All current Tasks now have pull requests associated with them, and \#8 is merged already.

**Comment by Zac-HD** [comment]
Added one last task based on a suggestion in the upstream issue 🙂

**Comment by Zac-HD** [comment]
All done 🎉
